File: backend/internal/models/money.go
// Package models holds the application's domain types. It has no knowledge
// of PocketBase or Square: conversion to and from storage happens in
// internal/database, and to and from Square in internal/providers.
package models

import (
	"fmt"
	"strconv"
)

// Currency is an ISO 4217 currency code, e.g. "USD".
type Currency string

const USD Currency = "USD"

// zeroDecimalCurrencies have no minor unit (1 JPY is the smallest amount).
var zeroDecimalCurrencies = map[Currency]bool{"JPY": true, "KRW": true, "VND": true, "CLP": true, "ISK": true}

// Money is an amount in the currency's smallest unit (cents for USD).
// Never use floating point for money.
type Money struct {
	Amount   int64
	Currency Currency
}

func NewMoney(amount int64, currency Currency) Money {
	return Money{Amount: amount, Currency: currency}
}

// Add returns m+o. Adding a zero-valued Money (no currency) is allowed so
// sums can start from Money{}.
func (m Money) Add(o Money) (Money, error) {
	switch {
	case m.Currency == "":
		return o, nil
	case o.Currency == "":
		return m, nil
	case m.Currency != o.Currency:
		return Money{}, fmt.Errorf("currency mismatch: %s + %s", m.Currency, o.Currency)
	}
	return Money{Amount: m.Amount + o.Amount, Currency: m.Currency}, nil
}

// Times returns m multiplied by a quantity.
func (m Money) Times(qty int64) Money {
	return Money{Amount: m.Amount * qty, Currency: m.Currency}
}

// String formats for display, e.g. "$8.50" or "1200 JPY".
func (m Money) String() string {
	sign := ""
	amount := m.Amount
	if amount < 0 {
		sign, amount = "-", -amount
	}
	if zeroDecimalCurrencies[m.Currency] {
		return sign + strconv.FormatInt(amount, 10) + " " + string(m.Currency)
	}
	major, minor := amount/100, amount%100
	if m.Currency == USD || m.Currency == "" {
		return fmt.Sprintf("%s$%d.%02d", sign, major, minor)
	}
	return fmt.Sprintf("%s%d.%02d %s", sign, major, minor, m.Currency)
}

File: app/Support/Money.php
<?php

declare(strict_types=1);

namespace App\Support;

use InvalidArgumentException;

/**
 * An amount in a currency's smallest unit (cents for USD). Money never uses
 * floating point: every price, subtotal, tax and total in this application is
 * an integer number of minor units.
 *
 * @phpstan-type MoneyArray array{cents: int, currency: string, formatted: string}
 */
final readonly class Money
{
    /** Currencies with no minor unit: 1 JPY is already the smallest amount. */
    private const ZERO_DECIMAL = ['JPY', 'KRW', 'VND', 'CLP', 'ISK'];

    public function __construct(
        public int $cents,
        public string $currency,
    ) {}

    public static function of(int $cents, ?string $currency = null): self
    {
        return new self($cents, $currency ?? self::defaultCurrency());
    }

    public static function zero(?string $currency = null): self
    {
        return self::of(0, $currency);
    }

    public static function defaultCurrency(): string
    {
        $currency = config('square.currency', 'USD');

        return is_string($currency) ? $currency : 'USD';
    }

    public function plus(self $other): self
    {
        $this->assertSameCurrency($other);

        return new self($this->cents + $other->cents, $this->currency);
    }

    public function times(int $quantity): self
    {
        return new self($this->cents * $quantity, $this->currency);
    }

    public function equals(self $other): bool
    {
        return $this->cents === $other->cents && $this->currency === $other->currency;
    }

    public function isZero(): bool
    {
        return $this->cents === 0;
    }

    /** Formats for display, e.g. "$8.50" or "1200 JPY". */
    public function format(): string
    {
        $sign = $this->cents < 0 ? '-' : '';
        $amount = abs($this->cents);

        if (in_array($this->currency, self::ZERO_DECIMAL, true)) {
            return "{$sign}{$amount} {$this->currency}";
        }

        $major = intdiv($amount, 100);
        $minor = str_pad((string) ($amount % 100), 2, '0', STR_PAD_LEFT);

        return $this->currency === 'USD'
            ? "{$sign}\${$major}.{$minor}"
            : "{$sign}{$major}.{$minor} {$this->currency}";
    }

    /**
     * The shape every Inertia page receives money in: the authoritative cents
     * plus a display string, so React never formats currency itself.
     *
     * @return MoneyArray
     */
    public function toArray(): array
    {
        return [
            'cents' => $this->cents,
            'currency' => $this->currency,
            'formatted' => $this->format(),
        ];
    }

    private function assertSameCurrency(self $other): void
    {
        if ($this->currency !== $other->currency) {
            throw new InvalidArgumentException(
                "Cannot combine {$this->currency} with {$other->currency}."
            );
        }
    }
}
